incubator/hnc/internal/reconcilers/object_test.go

```diff
@@ -15,6 +15,201 @@ import (
+			selector     string
+			treeSelector string
+			want         []string
+			notWant      []string
+		}{
+			{name: "simply using select",
+				selector:     "!" + c1 + api.LabelTreeDepthSuffix,
+				treeSelector: "",
+				want:         []string{c2, c3},
+				notWant:      []string{c1},
+			},
+			{name: "using multiple select",
+				selector:     "!" + c1 + api.LabelTreeDepthSuffix + ", !" + c2 + api.LabelTreeDepthSuffix,
+				treeSelector: "",
+				want:         []string{c3},
+				notWant:      []string{c1, c2},
+			},
+			{name: "simply using treeSelect",
+				selector:     "",
+				treeSelector: "!" + c1,
+				want:         []string{c2, c3},
+				notWant:      []string{c1},
+			},
+			{name: "using multiple treeSelect",
+				selector:     "",
+				treeSelector: "!" + c1 + ", !" + c2,
+				want:         []string{c3},
+				notWant:      []string{c1, c2},
+			},
+			{name: "using both select and treeSelect negatively",
+				selector:     "!" + c1 + api.LabelTreeDepthSuffix,
+				treeSelector: "!" + c2,
+				want:         []string{c3},
+				notWant:      []string{c1, c2},
+			},
+			{name: "using both select and treeSelect positively",
+				selector:     c1 + api.LabelTreeDepthSuffix,
+				treeSelector: c2,
+				want:         []string{},
+				notWant:      []string{c1, c2, c3},
+			},
+		}
+
+		for _, tc := range tests {
+			// Making a local copy of tc is necessary to ensure the correct value is passed to the closure,
+			// for more details look at https://onsi.github.io/ginkgo/ and search for 'closure'
+			tc := tc
+			It("Should handle exceptions: "+tc.name, func() {
+				// Set up namespaces
+				fooName := createNS(ctx, "foo")
+				barName := createNS(ctx, "bar")
+				bazName := createNS(ctx, "baz")
+				quxName := createNS(ctx, "qux")
+
+				setParent(ctx, barName, fooName)
+				setParent(ctx, bazName, fooName)
+				setParent(ctx, quxName, fooName)
+
+				// Convert namespace index to real names
+				indToNm := map[string]string{
+					p:  fooName,
+					c1: barName,
+					c2: bazName,
+					c3: quxName,
+				}
+				tc.selector = replaceStrings(tc.selector, indToNm)
+				tc.treeSelector = replaceStrings(tc.treeSelector, indToNm)
+
+				// Create a Role with the selector and treeSelector annotation
+				a := make(map[string]string)
+				if tc.selector != "" {
+					a["propagate.hnc.x-k8s.io/select"] = tc.selector
+				}
+				if tc.treeSelector != "" {
+					a["propagate.hnc.x-k8s.io/treeSelect"] = tc.treeSelector
+				}
+				makeObjectWithAnnotation(ctx, "Role", fooName, "testrole", a)
+				for _, ns := range tc.want {
+					ns = replaceStrings(ns, indToNm)
+					Eventually(hasObject(ctx, "Role", ns, "testrole")).Should(BeTrue(), "When propagating testrole to %s", ns)
+				}
+				for _, ns := range tc.notWant {
+					ns = replaceStrings(ns, indToNm)
+					Consistently(hasObject(ctx, "Role", ns, "testrole")).Should(BeFalse(), "When propagating testrole to %s", ns)
+				}
+			})
+		}
+	})
+
+	Context("Update exception annotations ", func() {
```

Comment:
What's the difference between the update cases with the add test cases?

Reply:
The update case checks that a ns had `testrole` but got removed after we update `testrole` with selectors, while the add case never updates the object, it was created with the selector so we only check each ns once